File: PDOO/QytetecJava/src/modeloqytetec/Qytetet.java

package modeloqytetec;

import java.util.ArrayList;
import java.util.Random;

/**
 * @author Jorge Valenzuela garcia
 */
public class Qytetet {
    
    // ATRIBUTOS
    private static final Qytetet instance = new Qytetet();
    public static final int MAX_JUGADORES = 4 ;
    protected static final int MAX_CARTAS = 10;
    protected static final int MAX_CASILLAS = 20;
    protected static final int PRECIO_LIBERTAD = 200;
    protected static final int SALDO_SALIDA = 1000;
    
    // ATRIBUTOS REFERENCIADOS
    //private Dado dado = Dado.getInstance();
    private Sorpresa cartaActual;
    private ArrayList<Sorpresa> mazo;
    private Jugador jugadorActual;
    private ArrayList<Jugador> jugadores;
    private Tablero tablero;
    
    // El constructor privado asegura que no se puede instanciar
    // desde otras clases
    private Qytetet(){}
    
    
    // MÉTODOS
    
    public boolean aplicarSorpresa(){
        boolean tienePropietario = false;
        TipoSorpresa tipo = this.cartaActual.getTipo();
        switch(tipo){
        
            case PAGARCOBRAR:
                this.jugadorActual.modificarSaldo(this.cartaActual.getValor());
                this.mazo.add(this.cartaActual);
                break;
            
                
            case IRACASILLA:
                int numeroCasilla = this.cartaActual.getValor();
                boolean esCarcel = this.tablero.esCasillaCarcel(numeroCasilla);
                if(esCarcel)
                    this.encarcelarJugador();
                else{
                    Casilla nuevaCasilla = this.tablero.obtenerCasillaNumero(numeroCasilla);
                    tienePropietario = this.jugadorActual.actualizarPosicion(nuevaCasilla);
                }
                this.mazo.add(this.cartaActual);
                break;
            
                
            case PORCASAHOTEL:
                this.jugadorActual.pagarCobrarPorCasaYHotel(this.cartaActual.getValor());
                this.mazo.add(this.cartaActual);
                break;
           
                
            case PORJUGADOR:
                for(Jugador jugador : this.jugadores){
                    // Si el "jugador" es distinto del jugadorActual...
                    if(!jugador.equals(this.jugadorActual)){
                        jugador.modificarSaldo(this.cartaActual.getValor());
                        this.jugadorActual.modificarSaldo(-this.cartaActual.getValor());
                    }
                }
                this.mazo.add(this.cartaActual);
                break;
        
                
            case SALIRCARCEL:
                this.jugadorActual.setCartaLibertad(this.cartaActual);
                break;

            case CONVERTIRME:
                // Cogemos el indice del jugador actual
                int indice = this.jugadores.indexOf(this.jugadorActual);
                // Creamos el objeto especulador del jugador actual
                Especulador espec = this.jugadorActual.convertirme(this.cartaActual.getValor());
                // Añadimos el especulador creado al array de jugadores en la posicion de jugador actual
                this.jugadores.add(indice, espec);
                // Volvemos a coger el nuevo indice de jugador actual
                indice = jugadores.indexOf(this.jugadorActual);
                // Y lo borramos
                jugadores.remove(indice);
                // Finalmente guardamos en jugador actual el nuevo especulador
                this.jugadorActual = espec;
                break;
        }
        
        return tienePropietario;
    }
    
    public boolean cancelarHipoteca(Casilla casilla){
        boolean cancelar = false;
        if(casilla.soyEdificable()){
            Calle calle = (Calle) casilla;
            if(calle.estaHipotecada()){
                if(this.jugadorActual.puedoPagarHipoteca(casilla)){
                    cancelar = true;
                    int cantidadRecibida = calle.cancelarHipoteca();
                    this.jugadorActual.modificarSaldo(-cantidadRecibida);
                }
            }
        }
        return cancelar;
    }
    
    public boolean comprarTituloPropiedad(){
        return this.jugadorActual.comprarTitulo();
    }
    
    public boolean edificarCasa(Casilla casilla){
        boolean puedoEdificar = false;
        if(casilla.soyEdificable()){
            Calle calle = (Calle) casilla;
            if(calle.sePuedeEdificarCasa(this.jugadorActual.getFactor())){
                if(this.jugadorActual.puedoEdificarCasa(casilla)){
                    int costeEdificarCasa = calle.edificarCasa();
                    this.jugadorActual.modificarSaldo(-costeEdificarCasa);
                    puedoEdificar = true;
                }
            }
        }
        return puedoEdificar;
    }
    
    public boolean edificarHotel(Casilla casilla){
        boolean puedoEdificar = false;
        if(casilla.soyEdificable()){
            Calle calle = (Calle) casilla;
            if(calle.sePuedeEdificarHotel(this.jugadorActual.getFactor())){
                if(this.jugadorActual.puedoEdificarHotel(casilla)){
                    int costeEdificarHotel = calle.edificarHotel();
                    this.jugadorActual.modificarSaldo(-costeEdificarHotel);
                    puedoEdificar = true;
                }
            }
        }
        return puedoEdificar;
    }
    
    public Sorpresa getCartaActual(){
        return this.cartaActual;
    }
    
    public static Qytetet getInstance(){
        return instance;
    }
    
    public Jugador getJugadorActual(){
        return this.jugadorActual;
    }
    
    public ArrayList<Jugador> getJugadores(){
        return this.jugadores;
    }
    
    public boolean hipotecarPropiedad(Casilla casilla){
        boolean puedoHipotecarPropiedad = false;
        Calle calle = (Calle) casilla;
        if(casilla.soyEdificable())
            if(!calle.estaHipotecada())
                if(this.jugadorActual.puedoHipotecar(casilla)){
                    puedoHipotecarPropiedad = true;
                    int cantidadRecibida = calle.hipotecar();
                    this.jugadorActual.modificarSaldo(cantidadRecibida);
                }
        return puedoHipotecarPropiedad;
    }
    
    public void inicializarJuego(ArrayList<String> nombres) throws Exception{
        this.inicializarJugadores(nombres); // Este metodo puede lanzar excepcion en caso nombres < 2
        this.inicializarCartasSorpresa();
        this.inicializarTablero();
        this.salidaJugadores();
    }
    
    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo){
        boolean libre = false;
        GUIQytetet.Dado dado = GUIQytetet.Dado.getInstance();
        if(metodo == MetodoSalirCarcel.TIRANDODADO){
            int valorDado = dado.nextNumber();
            System.out.println("Ha salido: " + Integer.toString(valorDado));
            libre = valorDado > 5;
        }
        else{
            boolean tengoSaldo = this.jugadorActual.pagarLibertad(Qytetet.PRECIO_LIBERTAD);
            if (tengoSaldo)
                libre = true;
        }
        if(libre)
            this.jugadorActual.setEncarcelado(false);
        return libre;
    }
    
    public boolean jugar(){
        GUIQytetet.Dado dado = GUIQytetet.Dado.getInstance();
        int valorDado = dado.nextNumber();
        System.out.println(valorDado);
        Casilla casillaPosicion = this.jugadorActual.getCasillaActual();
        Casilla nuevaCasilla = this.tablero.obtenerNuevaCasilla(casillaPosicion, valorDado);
        boolean tienePropietario = this.jugadorActual.actualizarPosicion(nuevaCasilla);
        if(!nuevaCasilla.soyEdificable()){
            if(nuevaCasilla.getTipo().equals(TipoCasilla.JUEZ))
                this.encarcelarJugador();
            else{
                if(nuevaCasilla.getTipo().equals(TipoCasilla.SORPRESA)){
                    // Cogemos la primera carta del mazo
                    this.cartaActual = this.mazo.remove(0);
                }
            }
        }
        return tienePropietario;
    }
    
    public ArrayList obtenerRanking(){//
        return null;
    }
    
    public ArrayList<Casilla> propiedadesHipotecadasJugador(boolean hipotecadas){
        ArrayList<Casilla> devolver = new ArrayList();
        ArrayList<TituloPropiedad> listaPropiedades = this.jugadorActual.obtenerPropiedadesHipotecadas(hipotecadas);
        for (TituloPropiedad titulo : listaPropiedades){
            devolver.add(titulo.getCasilla());
        }
        return devolver;
    }
    
    public void siguienteJugador(){
        int numero_jugadores = this.jugadores.size();
        int numero_jugador = this.jugadores.indexOf(this.jugadorActual);
        if (numero_jugador == numero_jugadores -1) // Si es el ultimo jugador
            this.jugadorActual = this.jugadores.get(0);
        else
            this.jugadorActual = this.jugadores.get(numero_jugador+1);
    }
    
    public boolean venderPropiedad(Casilla casilla){
        boolean puedoVender = false;
        if(casilla.soyEdificable()){
            puedoVender = this.jugadorActual.puedoVenderPropiedad(casilla);
            if(puedoVender){
                this.jugadorActual.venderPropiedad(casilla);
            }
        }
        return puedoVender;
    }
    
    private void encarcelarJugador(){
        if(!this.jugadorActual.tengoCartaLibertad()){
            Casilla carcel = this.tablero.getCarcel();
            this.jugadorActual.irACarcel(carcel);
        }
        else{
            Sorpresa carta = this.jugadorActual.devolverCartaLibertad();
            this.mazo.add(carta);
        }
    }
    
    private void inicializarCartasSorpresa(){
        this.mazo = new ArrayList();
        this.mazo.add( new Sorpresa ("Parece que te vas a convertir en un especulador...", 3000, TipoSorpresa.CONVERTIRME));
        this.mazo.add( new Sorpresa ("Te han dado una donación de 150 euros", 150, TipoSorpresa.PAGARCOBRAR));    
        this.mazo.add( new Sorpresa ("Te has dejado el coche donde no debías, pagas una multa de 150 euros", -150, TipoSorpresa.PAGARCOBRAR));      
        this.mazo.add( new Sorpresa ("Te hemos pillado con chanclas y calcetines, lo sentimos, ¡debes ir a la carcel!", 11, TipoSorpresa.IRACASILLA));
        this.mazo.add( new Sorpresa ("La patrulla antisuperstición te lleva a la casilla 13", 13, TipoSorpresa.IRACASILLA));
        this.mazo.add( new Sorpresa ("Se te ve con ganas de volver a empezar, vuelve a la casilla de salida", 0, TipoSorpresa.IRACASILLA));
        this.mazo.add( new Sorpresa ("Inspección te ha pillado, pagas 50€ por cada propiedad que tengas", -50, TipoSorpresa.PORCASAHOTEL));
        this.mazo.add( new Sorpresa ("Has generado muchos beneficios, por cada propiedad cobras 50 euros", 50, TipoSorpresa.PORCASAHOTEL));
        this.mazo.add( new Sorpresa ("Es tu cumpleaños, cada jugador te da 20 euros", -20, TipoSorpresa.PORJUGADOR));
        this.mazo.add( new Sorpresa ("Has perdido una apuesta, debes darle a cada jugador 20 euros", 20, TipoSorpresa.PORJUGADOR));
        this.mazo.add( new Sorpresa ("Un fan anónimo ha pagado tu fianza. Sales de la cárcel", 0, TipoSorpresa.SALIRCARCEL));
        this.mazo.add( new Sorpresa ("Parece que te vas a convertir en un especulador...", 5000, TipoSorpresa.CONVERTIRME));
    }
    
    private void inicializarJugadores(ArrayList<String> nombres) throws Exception{
        this.jugadores = new ArrayList();
        if (nombres.size() >= 2 && nombres.size() <= MAX_JUGADORES){
            for(String nombre : nombres){
                Jugador jugador = new Jugador(nombre);
                this.jugadores.add(jugador);
            }
        }
        else{
            throw new Exception("Numero de jugadores incorrecto. Para jugar se necesitan dos o mas jugadores");
        }
    }
    
    private void inicializarTablero(){
            this.tablero = new Tablero();
    }
    
    private void salidaJugadores(){
        for (Jugador jugador : this.jugadores){
            jugador.setCasillaActual(this.tablero.obtenerCasillaNumero(0));
        }
        Random randomGenerator = new Random();
        int posicion = (randomGenerator.nextInt() % this.jugadores.size());
        if(posicion < 0)
            posicion = -posicion;
        this.jugadorActual = this.jugadores.get(posicion);
    }
    
    
    @Override
    public String toString(){
        String salida;
        salida =    "Qytetet:" +
                    "\nMaximo de jugadores: " + Integer.toString(MAX_JUGADORES) + 
                    "\nMaximo de cartas: " + Integer.toString(MAX_CARTAS) +
                    "\nMaximo de casillas: " + Integer.toString(MAX_CASILLAS) +
                    "\nPrecio de libertad: " + Integer.toString(PRECIO_LIBERTAD) +
                    "\nSaldo de salida: " + Integer.toString(SALDO_SALIDA);
        //salida += "\nAun faltan datos para una correcta inicializacion de los atributos de instancia";
        return salida;
    }
    
}

File: PDOO/QytetetRuby/pkg/QytetetRuby-0.0.1/lib/qytetet.rb
#encoding: utf-8

require "singleton"
require_relative "metodo_salir_carcel"
require_relative "dado"
require_relative "tablero"
require_relative "casilla"
require_relative "jugador"
require_relative "sorpresa"
require_relative "tipo_sorpresa"
require_relative "tipo_casilla"
require_relative "titulo_propiedad"

module ModeloQytetet
  class Qytetet
    include Singleton
    
      @@MAX_JUGADORES = 4
      @@MAX_CARTAS = 10
      @@MAX_CASILLAS = 20
      @@PRECIO_LIBERTAD = 2000
      @@SALDO_SALIDA = 1000
    
    def initialize
      @dado = Dado.instance
      @carta_actual
      @jugador_actual
      @tablero
      @mazo
      @jugadores
    end
    
    attr_reader :carta_actual, :jugador_actual, :SALDO_SALIDA
    
    def to_s
      #puts "Metodo aun no implementado debido a la falta de datos para una correcta inicializacion"
      puts "Maximo de jugadores: #{@@MAX_JUGADORES}"
      puts "Maximo de cartas: #{@@MAX_CARTAS}"
      puts "Maximo de casillas: #{@@MAX_CASILLAS}"
      puts "Precio de libertad: #{@@PRECIO_LIBERTAD}"
      puts "Saldo de salida: #{@@SALDO_SALIDA}"
      # puts "Aun faltan los atributos de instancia, que por falta de datos no se pueden terminar de inicilizar correctamente"
    end
    
    
    def aplicar_sorpresa
      tiene_propietario = false
      puts @carta_actual.tipo.to_s
      tipo_carta = @carta_actual.tipo
      case tipo_carta        
      when TipoSorpresa::PAGARCOBRAR
        @jugador_actual.modificar_saldo(@carta_actual.valor)
        @mazo << @carta_actual
        
      when TipoSorpresa::IRACASILLA
        numero_casilla = @carta_actual.valor
        es_carcel = @tablero.es_casilla_carcel(numero_casilla)
        if es_carcel
          encarcelar_jugador
        else
          nueva_casilla = @tablero.obtener_casilla_numero(numero_casilla)
          tiene_propietario = @jugador_actual.actualizar_posicion(nueva_casilla)
        end
        @mazo << @carta_actual
        
      when TipoSorpresa::PORCASAHOTEL
        @jugador_actual.pagar_cobrar_por_casa_y_hotel(@carta_actual.valor)
        @mazo << @carta_actual
        
      when TipoSorpresa::PORJUGADOR
        @jugadores.each do |jug|
          if !jug.eql?(@jugador_actual)
            jug.modificar_saldo(@carta_actual.valor)
            @jugador_actual.modificar_saldo(-@carta_actual.valor)
          end
        end
        @mazo << @carta_actual

      when TipoSorpresa::SALIRCARCEL
        @jugador_actual.carta_libertad = @carta_actual
      end

      return tiene_propietario
    end
    
    
    
    def cancelar_hipoteca(casilla)
      cancelar = false
      if casilla.soy_edificable
        if casilla.esta_hipotecada
          if @jugador_actual.puedo_pagar_hipoteca(casilla)
            cancelar = true
            cantidad_recibida = casilla.cancelar_hipoteca
            @jugador_actual.modificar_saldo(-cantidad_recibida)
          end
        end
      end
      return cancelar
    end
    
    
    
    def comprar_titulo_propiedad()
      @jugador_actual.comprar_titulo
    end
    
    
    
    def edificar_casa(casilla)
      puedo_edificar = false
      if casilla.soy_edificable
        if casilla.se_puede_edificar_casa
          if @jugador_actual.puedo_edificar_casa(casilla)
            coste_edificar_casa = casilla.edificar_casa
            @jugador_actual.modificar_saldo(-coste_edificar_casa)
            puedo_edificar = true
          end
        end
      end
      
      return puedo_edificar
    end
    
    
    
    def edificar_hotel(casilla)
      puedo_edificar = false
      if casilla.soy_edificable
        if casilla.se_puede_edificar_hotel
          if @jugador_actual.puedo_edificar_hotel(casilla)
            coste_edificar_hotel = casilla.edificar_hotel
            @jugador_actual.modificar_saldo(-coste_edificar_hotel)
            puedo_edificar = true
          end
        end
      end
      
      return puedo_edificar
    end
    
    
    
    def hipotecar_propiedad(casilla)
      puedo_hipotecar_propiedad = false
      if casilla.soy_edificable
        if !casilla.esta_hipotecada
          if @jugador_actual.puedo_hipotecar(casilla)
            puedo_hipotecar_propiedad = true
            cantidad_recibida = casilla.hipotecar
            @jugador_actual.modificar_saldo(cantidad_recibida)
          end
        end
      end
      
      return puedo_hipotecar_propiedad
    end
    
    
    
    def inicializar_juego(nombres)
      inicializar_jugadores(nombres)
      inicializar_cartas_sorpresa
      inicializar_tablero
      salida_jugadores
    end
    
    
    
    def intentar_salir_carcel(metodo)
      libre = false
      
      if metodo == MetodoSalirCarcel::TIRANDODADO
        valor_dado = @dado.tirar
        if valor_dado > 5
          libre = true
        end
      else
        tengo_saldo = @jugador_actual.pagar_libertad(@@PRECIO_LIBERTAD)
        if tengo_saldo
          libre = true
        end
      end
      
      if libre
        @jugador_actual.encarcelado = false
      end
      return libre
    end
    
    
    
    def jugar()
      valor_dado = @dado.tirar
      casilla_posicion = @jugador_actual.casilla_actual
      nueva_casilla = @tablero.obtener_nueva_casilla(casilla_posicion, valor_dado)
      tiene_propietario = @jugador_actual.actualizar_posicion(nueva_casilla)
      if !nueva_casilla.soy_edificable
        if nueva_casilla.tipo.eql?(TipoCasilla::JUEZ)
          encarcelar_jugador
        else
          if nueva_casilla.tipo.eql?(TipoCasilla::SORPRESA)
            @carta_actual = @mazo[0]
            @mazo.shift
          end
        end
      end
      
      return tiene_propietario
    end
    
    
    
    def obtener_ranking ## Aun sin implementar
      puts "No implementado"
    end
    
    
    
    def propiedades_hipotecadas_jugador(hipotecadas)  ### Posible error
      devolver = Array.new
      lista_propiedades = @jugador_actual.obtener_propiedades_hipotecadas(hipotecadas)
      lista_propiedades.each { |propiedad| 
        devolver.push(propiedad.casilla)}
      return devolver
    end
    
    
    
    def siguiente_jugador ### Posible error
      numero_jugadores = @jugadores.size
      numero_jugador = @jugadores.index(@jugador_actual)
      if (numero_jugador == numero_jugadores -1) ## Si es el ultimo jugador
        @jugador_actual = @jugadores[0]
      else
        @jugador_actual = @jugadores[numero_jugador +1]
      end  
    end
    
    
    
    def vender_propiedad(casilla)
      puedo_vender = false
      if casilla.soy_edificable
        puedo_vender = @jugador_actual.puedo_vender_propiedad(casilla)
        if puedo_vender
          @jugador_actual.vender_propiedad(casilla)
        end
      end
      return puedo_vender
    end
    
    
    
    
    private
    def encarcelar_jugador
      if !@jugador_actual.tengo_carta_libertad()
        carcel = @tablero.carcel
        @jugador_actual.ir_a_carcel(carcel)
      else
        carta = @jugador_actual.devolver_carta_libertad
        @mazo << carta
      end
    end
    
    
    
    def inicializar_cartas_sorpresa 
      @mazo = Array.new
      @mazo << Sorpresa.new("Te han dado una donacion de 150 monedas de oro", 150, TipoSorpresa::PAGARCOBRAR)
      @mazo << Sorpresa.new("Te has dejado el carruaje donde no debias, pagas una multa de 150 monedas de oro", -150, TipoSorpresa::PAGARCOBRAR)
      @mazo << Sorpresa.new("Te hemos pillado con chanclas y calcetines, lo sentimos, debes ir a Mordor!", 11, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("La patrulla antisuperstición te lleva a la casilla 13", 13, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("Se te ve con ganas de volver a empezar, vuelve a la casilla de salida", 0, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("Inspección te ha pillado, pagas 50€ por cada propiedad que tengas", -50, TipoSorpresa::PORCASAHOTEL)
      @mazo << Sorpresa.new("Has generado muchos beneficios, por cada propiedad cobras 50 monedas de oro", 50, TipoSorpresa::PORCASAHOTEL)
      @mazo << Sorpresa.new("Es tu cumpleaños, cada jugador te da 20 monedas de oro", 20, TipoSorpresa::PORJUGADOR)
      @mazo << Sorpresa.new("Has perdido una apuesta, debes darle a cada jugador 20 monedas de oro", -20, TipoSorpresa::PORJUGADOR)
      @mazo << Sorpresa.new("Escapas de Mordor", 0, TipoSorpresa::SALIRCARCEL)
    end
      
    
    
    def inicializar_jugadores(nombres)
      if nombres.size >= 2 and nombres.size <= @@MAX_JUGADORES
        @jugadores = Array.new
        nombres.each do |nombre|
          @jugadores << Jugador.new(nombre)
        end
      else
        puts "Los jugadores tienen que estar en entre 2 y #{@@MAX_JUGADORES} "
      end
    end
    
    
    
    def inicializar_tablero
      @tablero = Tablero.new
    end
    
    
    
    def salida_jugadores ### Posible error
      @jugadores.each do |jug|  ## Ponemos a todos los jugadores en la casilla 0
        jug.casilla_actual = @tablero.obtener_casilla_numero(0)
      end
      posicion = rand(0..@jugadores.size-1)
      @jugador_actual = @jugadores[posicion.to_int]
    end
    
    
    
  end
end
